crates/net/ress/src/connection.rs

```diff
@@ -0,0 +1,343 @@
+    fn next_id(&mut self) -> u64 {
+        let id = self.next_id;
+        self.next_id += 1;
+        id
+    }
+
+    /// Report bad message from current peer.
+    fn report_bad_message(&self) {
+        self.peers_handle.reputation_change(self.peer_id, ReputationChangeKind::BadMessage);
+    }
+
+    fn on_command(&mut self, command: RessPeerRequest) -> RessProtocolMessage {
+        let next_id = self.next_id();
+        let message = match &command {
+            RessPeerRequest::GetHeaders { request, .. } => {
+                RessProtocolMessage::get_headers(next_id, *request)
+            }
+            RessPeerRequest::GetBlockBodies { request, .. } => {
+                RessProtocolMessage::get_block_bodies(next_id, request.clone())
+            }
+            RessPeerRequest::GetWitness { block_hash, .. } => {
+                RessProtocolMessage::get_witness(next_id, *block_hash)
+            }
+            RessPeerRequest::GetBytecode { code_hash, .. } => {
+                RessProtocolMessage::get_bytecode(next_id, *code_hash)
+            }
+        };
+        self.inflight_requests.insert(next_id, command);
+        message
+    }
+}
+
+impl<P> RessProtocolConnection<P>
+where
+    P: RessProtocolProvider + Clone + 'static,
+{
+    fn on_headers_request(&self, request: GetHeaders) -> Vec<Header> {
+        match self.provider.headers(request) {
+            Ok(headers) => headers,
+            Err(error) => {
+                trace!(target: "ress::net::connection", peer_id = %self.peer_id, ?request, %error, "error retrieving headers");
+                Default::default()
+            }
+        }
+    }
+
+    fn on_block_bodies_request(&self, request: Vec<B256>) -> Vec<BlockBody> {
+        match self.provider.block_bodies(request.clone()) {
+            Ok(bodies) => bodies,
+            Err(error) => {
+                trace!(target: "ress::net::connection", peer_id = %self.peer_id, ?request, %error, "error retrieving block bodies");
+                Default::default()
+            }
+        }
+    }
+
+    fn on_bytecode_request(&self, code_hash: B256) -> Bytes {
+        match self.provider.bytecode(code_hash) {
+            Ok(Some(bytecode)) => bytecode,
+            Ok(None) => {
+                trace!(target: "ress::net::connection", peer_id = %self.peer_id, %code_hash, "bytecode not found");
+                Default::default()
+            }
+            Err(error) => {
+                trace!(target: "ress::net::connection", peer_id = %self.peer_id, %code_hash, %error, "error retrieving bytecode");
+                Default::default()
+            }
+        }
+    }
+
+    fn on_witness_request(&self, request: RequestPair<B256>) {
+        let peer_id = self.peer_id;
+        let provider = self.provider.clone();
+        let (tx, rx) = oneshot::channel();
+        tokio::spawn(async move {
+            let block_hash = request.message;
+            let witness = match provider.witness(block_hash).await {
```

Comment:
why is this a future, if this is already a future then we don't need to spawn it? and can store it directly,

there's no reference impl so unsure how this should be implemented.